better_words/message_parser.py

```diff
@@ -0,0 +1,66 @@
+from os import listdir
+from os.path import splitext
+from collections import namedtuple
+from db import session
+from models import *
+import re
+import yaml
+from sqlalchemy import and_
+
+# A named tuple representing a match from the parser
+ParserMatch = namedtuple('ParserMatch', 'message model word')
+FormattedMatch = namedtuple('FormattedMatch', 'intro explanation')
+
+# TODO: handle messages that contain multiple matches
+# TODO: check if the user has opted out
+def is_match(message):
+    """
+    Check if the given message contains a word or phrase match
+
+    @param message: the message to check
+    @return: a ParserMatch tuple if a match is found, otherwise None
+    """
+    user_id = message.get('user')
+
+    # Fetch the user from the DB, or add if needed
+    user = session.query(User).get(user_id)
+    if not user:
+        user = session.add(User(id=user_id))
+        session.commit()
+
+    # Find all the words the users hasn't opted out of
+    words = session.query(Word)\
+        .outerjoin(UserCategoryOptOut,
+            and_(UserCategoryOptOut.category_slug == Word.category_slug,
+                UserCategoryOptOut.user_id == user.id))\
+        .filter(UserCategoryOptOut.category_slug == None)
```

Comment:
This results in a query like (since the whole sqlalchemy syntax was initially a messy mystery for me)
```sql
SELECT * FROM words 
LEFT JOIN user_category_opt_out as coo
  ON coo.category_slug == words.category_slug
  AND coo.user_id == ?
WHERE coo.category_slug IS NOT NULL
```